packages/event-bus-redis/src/services/event-bus-redis.ts

```diff
@@ -53,86 +49,64 @@ type EmitOptions = {
  * Can keep track of multiple subscribers to different events and run the
  * subscribers when events happen. Events will run asynchronously.
  */
-export default class EventBusService {
+export default class RedisEventBusService extends AbstractEventBusService {
   protected readonly config_: ConfigModule
-  protected readonly manager_: EntityManager
+  protected readonly moduleOptions_: ModuleOptions
+  protected readonly moduleDeclaration_: ConfigurableModuleDeclaration
   protected readonly logger_: Logger
-  protected readonly stagedJobRepository_: typeof StagedJobRepository
-  protected readonly jobSchedulerService_: JobSchedulerService
-  protected readonly eventToSubscribersMap_: Map<
-    string | symbol,
-    SubscriberDescriptor[]
-  >
-  protected readonly redisClient_: Redis.Redis
-  protected readonly redisSubscriber_: Redis.Redis
-  protected queue_: Bull
+  protected readonly stagedJobService_: StagedJobService
```

Comment:
thought (non-blocking): would it make sense to keep the outbox functionality internal to the core and instead only use the "EventBusModule" as a facilitator of adding messages to a queue/broker/stream/etc.?

In practice, this would mean:

- Core would have an "Event Bus wrapper" which depends on the installed module.
- Core takes care of adding events to StagedJob (if modules want to do the same, they would also do so internally)
- Core EventBus enqueuer makes the calls to [EventBusModule].emit
- If Core EventBus doesn't have a transaction manager we can just passthrough directly to the installed module.

This limits the scope of what an EventBusModule does to just focus on "take this data and put it into this system in the correct way". 

The core and each module can then choose to add additional guarantees, like an outbox pattern, but it is not being forced by the fact that you have installed a specific event bus module. 

-- 

This issue I am seeing with this one for example is that it relies on the consumer of this module to also share a connection to the DB so it can transactionally stage the events. But in a scenario where a module (not an EvenBusModule, but for example an inventory module) is isolated/standalone we might not have this connection available, and in such case the module is better off handling this stuff by itself.

I see how we could potentially just move this part to a second iteration, but I don't see a lot of value in having the outbox logic sit alongside the 3rd party (e.g., Redis) connection and communication logic - I imagine this would make it easier to build new connections + make it simpler to do the next iteration on standalone modules without having to revisit all of this.

Reply:
I agree that the outbox, staged jobs pattern could be part of the core, since the purpose is really to send the events at the right moment and not in a middle of a flow that is not finished and could be rollback. This would guarantee that you can trust the core to only send the events when they can be handled and not before.

I quite like the idea of having a service, on front of the module, in the core which act as a relay to the event module such as what we already have for the different providers. In that case, each module wanting to emit an event would call this service which in returns will call the underlying module. Therefore the connection is not an issue anymore, Have I understand correctly @srindom ?

Reply:
Great input, but from what I understand, this is broadly the discussion around Outbox with a Relay service vs. Cached events pattern that we are yet to decide on. Please correct me if I am wrong.

This is why the scope of this PR was originally constrained to only extracting the core event bus into a module and cleaning up the event bus service. 

The second iteration will be run when we've aligned on how to manage events across modules. If we choose the caching pattern, I expect we will remove the staged jobs concept entirely. And I can't see how the "relay service" would benefit this pattern, as we would then bound the context of event bus modules to _adding messages to a queue/broker/steam_, so we will only need an interface in the core to rely on. With that said, I'd rather hold back with that implementation, for now, to avoid building something that will be redundant near term. 

I am 100% aligned with you guys; all modules should be capable of running in isolation from the core.

Reply:
Another way to look at it: regardless of how we end up handling guarantees across modules when we get there - we will still need logic that can take the message from internal data and then put it into X event system. So if we separate those two things we avoid coupling a guarantee pattern (i.e., outbox/cache) with the logic to put the data into the provider.

Reply:
Yes, makes sense. The counter-point I was trying to make was that transactional guarantees in the core would not be solved via an outbox if we choose to go with the Cached Events pattern - unless we make it a priority to support both, which I think should be well thought through as this would mean maintaining two patterns in the core going forward. 

That said, I might need to give this some extra thought. Adding the relay service now is not super tedious or complex, but it would make us take the first step toward the outbox pattern, I believe.

Reply:
> we will still need logic that can take the message from internal data and then put it into X event system

I would imagine this is what the interface is for. Can you give me an example with the Redis event bus? I think that would help me understand.

Reply:
We already have an outbox pattern in the core so we don't really change anything - we just only extract the part of the event bus that "connects to the underlying provider". 

Will try to hash out very roughly here:

```js
// packages/medusa/src/services/event-bus.ts

class ... {
  constructor({ eventBusModule, stagedJob, ... }) {
    this.provider: IEventBusProvider = eventBusModule // This is the module that handles connection to Redis
    ...
  }

  emit(...) {
    if (this.transactionManager_) {
      // Add it to staged job
    } else {
      this.provider.emit(...)
    } 
  }
  
  enqueuer_() {
    // pick up from staged job
    this.provider.emit(dataFromStagedJob)
  }
}
```

---

```ts
// packages/event-bus-redis/src/services/event-bus.ts

class ... implements IEventBusProvider {
  // assume that we have connected to redis, configured bull, etc, etc.

  emit (...) {
    this.queue_.add(...)
  }
}

```

Reply:
@srindom Got it - this is what I imagined as well.

Can I get you to verify, that [this is](https://github.com/medusajs/medusa/pull/2599/commits/669f8af6cff2495f1d4d2d834e9168dfe327837a) the pattern that we could introduce?

Please note, this is just a quick first draft to ensure, that we are aligned.

Reply:
Another thought; would it make sense to always use the staged jobs mechanism to ensure complete resiliency against event loss that could happen on `.emit( ... )`? That gives us an at-least-once delivery guarantee. 

In combination with an improved and configurable polling mechanism, events should not be delayed much from this.

Reply:
We could change `emit` and `enqueuer` to work like so (in pseudo):
```
class EventRelayService
...
this.enqueue = false

emit {
  stagedJobService.add(job)
  this.enqueue = true
}

enqueuer {
  while (this.enqueue) {
    stagedJobService.list( ... )
    sleepWithExponentialBackoff( ... )

    if (currentAttempt === maxretries) {
      this.enqueue = false
    }
  }
} 

```

Reply:
Thanks for pushing this idea through. Have done the changes locally, and it definitely reduces the responsibility of the event bus module. Will push as soon as tests have been created.